lib/nx_signal/peak_finding.ex

```diff
@@ -0,0 +1,392 @@
+  a number is a relative maximum if it is greater than or equal to the double of its
+  neighbors, as follows:
+
+      iex> comparator = fn x, y -> Nx.greater_equal(x, Nx.multiply(y, 2)) end
+      iex> x = Nx.tensor([0, 1, 3, 2, 0, 1, 0, 0, 0, 2, 1])
+      iex> result = NxSignal.PeakFinding.argrelextrema(x, comparator)
+      iex> result.valid_indices
+      #Nx.Tensor<
+        u64
+        3
+      >
+      iex> result.indices[0..2]
+      #Nx.Tensor<
+        s64[3][1]
+        [
+          [5],
+          [7],
+          [9]
+        ]
+      >
+
+  Same applies for finding local minima. In the next example, we
+  find all local minima (i.e. `&Nx.less/2`) that are
+  different to the global minimum.
+
+      iex> x = Nx.tensor([0, 1, 0, 2, 1, 3, 0, 1])
+      iex> global_minimum = Nx.reduce_min(x)
+      iex> comparator = fn x, y ->
+      ...> x_not_global = Nx.not_equal(x, global_minimum)
+      ...> y_not_global = Nx.not_equal(y, global_minimum)
+      ...> both_not_global = Nx.logical_and(x_not_global, y_not_global)
+      ...> Nx.logical_and(Nx.less(x, y), both_not_global)
+      ...> end
+      iex> result = NxSignal.PeakFinding.argrelextrema(x, comparator)
+      iex> result.valid_indices
+      #Nx.Tensor<
+        u64
+        1
+      >
+      iex> result.indices[0..0]
+      #Nx.Tensor<
+        s64[1][1]
+        [
+          [4]
+        ]
+      >
+  """
+  @doc type: :peak_finding
+  defn argrelextrema(data, comparator_fn, opts \\ []) do
+    opts = keyword!(opts, axis: 0, order: 1)
+
+    data
+    |> boolrelextrema(comparator_fn, opts)
+    |> nonzero()
+  end
+
+  defnp boolrelextrema(data, comparator_fn, opts \\ []) do
+    axis = opts[:axis]
+    order = opts[:order]
+    locs = Nx.iota({Nx.axis_size(data, axis)})
+
+    ones = Nx.broadcast(u8(1), data.shape)
+    [ones, _] = Nx.broadcast_vectors([ones, data])
+
+    {results, _} =
+      while {results = ones, {data, locs, halt = u8(0), shift = s64(1)}},
+            not halt and shift < order + 1 do
+        plus = Nx.take(data, Nx.clip(locs + shift, 0, Nx.size(locs) - 1), axis: axis)
+        minus = Nx.take(data, Nx.clip(locs - shift, 0, Nx.size(locs) - 1), axis: axis)
+        results = comparator_fn.(data, plus) and results
+        results = comparator_fn.(data, minus) and results
+
+        {results, {data, locs, not Nx.any(results), shift + 1}}
+      end
+
+    results
+  end
+
+  deftransformp nonzero(data) do
+    flat_data = Nx.reshape(data, {:auto, 1})
+
+    indices =
+      for axis <- 0..(Nx.rank(data) - 1),
+          reduce: Nx.broadcast(0, {Nx.axis_size(flat_data, 0), Nx.rank(data)}) do
+        %{shape: {n, _}} = indices ->
+          iota = data.shape |> Nx.iota(axis: axis) |> Nx.reshape({n, 1})
+          Nx.put_slice(indices, [0, axis], iota)
+      end
+
+    indices_with_mask =
+      Nx.select(
+        Nx.broadcast(flat_data, indices.shape),
+        indices,
+        Nx.broadcast(-1, indices.shape)
```

Comment:
```suggestion
        Nx.broadcast(-1, indices)
```

Reply:
I prefer using shape for consistency with iota, which only accepts shape